NewHorizons/Patches/EchoesOfTheEyePatches/DreamworldControllerPatches.cs

```diff
@@ -26,5 +27,97 @@ public static IEnumerable<CodeInstruction> DreamWorldController_SpawnInDreamworl
+                var arrivalAO = arrivalPoint.GetComponentInParent<AstroObject>();
+                __instance._dreamBody = arrivalAO.GetAttachedOWRigidbody();
+                __instance._dreamWorldSector = arrivalAO.GetRootSector();
+                __instance._dreamWorldVolume = arrivalAO._gravityVolume.GetOWTriggerVolume();
+            }
+
+            // Make the 'bubble' around the artifact load correctly when the destination body isn't the vanilla DreamWorld
+            __instance._primarySimulationRoot.GetComponent<SectorCullGroup>().SetSector(__instance._dreamWorldSector);
+
+            // if the player's already in the dream world, we want to wake up at the first campfire we slept at, so don't override those values
+            if (PlayerState.InDreamWorld())
+            {
+                __instance._dreamArrivalPoint = arrivalPoint;
+                __instance._enteringDream = true;
+                return false;
+            }
+
+            // Make the body where the player 'wakes up' out of the dream match the body where the campfire is located if it isn't the Stranger ("RingWorld"), or reset it if it is
+            var ringWorldAO = Locator.GetAstroObject(AstroObject.Name.RingWorld);
+            if (dreamCampfire.GetAttachedOWRigidbody() == ringWorldAO.GetOWRigidbody())
+            {
+                __instance._planetBody = ringWorldAO.GetAttachedOWRigidbody();
+                __instance._ringWorldController = ringWorldAO.GetComponent<RingWorldController>();
+            }
+            else
+            {
+                var departureAO = dreamCampfire.GetComponentInParent<AstroObject>();
+                __instance._planetBody = departureAO.GetAttachedOWRigidbody();
+                __instance._ringWorldController = null;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Load assets when dying in dreamworld and returning to a custom dream campfire
+        /// Only really relevant for campfires on stock planets
+        /// 
+        /// This only gets us like a 1 second headstart on loading so honestly it's barely worth it,
+        /// could be replaced with keeping the assets loaded the entire time you're in the dreamworld
+        /// but we should only do that if it becomes relevant I think
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="dreamCampfire"></param>
+        [HarmonyPrefix, HarmonyPatch(nameof(DreamWorldController.EnterDreamWorld))]
+        private static void OnEnterDreamWorld(DreamWorldController __instance, DreamCampfire dreamCampfire)
+        {
+            var planet = dreamCampfire.GetComponentInParent<AstroObject>();
+            if (planet.GetAstroObjectName() != AstroObject.Name.RingWorld)
+            {
+                _dreamEntryLocationStreamingGroup = planet.GetComponentInChildren<StreamingGroup>();
+                GlobalMessenger<DeathType>.AddListener("PlayerDeath", OnPlayerDeath);
```

Comment:
The AddListener and RemoveListener calls could potentially end up mismatched if the dreamworld is entered recursively. If this logic was moved to after the early `return false;` in the patch above, that should avoid that issue.